test/util/url.test.cpp

```diff
@@ -23,3 +23,184 @@ TEST(URL, isURL) {
+
+TEST(Path, Directory) {
+    EXPECT_EQ(Path::Segment({ 0, 8 }), Path("foo/bar/baz.ext").directory);
+    EXPECT_EQ(Path::Segment({ 0, 8 }), Path("foo.bar/baz.ext").directory);
+    EXPECT_EQ(Path::Segment({ 0, 8 }), Path("foo.bar/baz").directory);
+    EXPECT_EQ(Path::Segment({ 0, 8 }), Path("foo/bar/.ext").directory);
+    EXPECT_EQ(Path::Segment({ 0, 4 }), Path("foo/bar@2x.png").directory);
+    EXPECT_EQ(Path::Segment({ 0, 4 }), Path("foo/b").directory);
+    EXPECT_EQ(Path::Segment({ 0, 4 }), Path("foo/").directory);
+    EXPECT_EQ(Path::Segment({ 0, 0 }), Path("foo").directory);
+    EXPECT_EQ(Path::Segment({ 0, 0 }), Path("foo.png").directory);
+    EXPECT_EQ(Path::Segment({ 0, 0 }), Path("").directory);
+}
+
+TEST(Path, URLDirectory) {
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo/bar/baz.ext").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo/bar/baz.ext?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo.bar/baz.ext").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo.bar/baz.ext?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo.bar/baz").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo.bar/baz?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo/bar/.ext").directory);
+    EXPECT_EQ(Path::Segment({ 19, 8 }), URLPath("http://example.com/foo/bar/.ext?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 4 }), URLPath("http://example.com/foo/bar@2x.png").directory);
+    EXPECT_EQ(Path::Segment({ 19, 4 }), URLPath("http://example.com/foo/bar@2x.png?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 4 }), URLPath("http://example.com/foo/b").directory);
+    EXPECT_EQ(Path::Segment({ 19, 4 }), URLPath("http://example.com/foo/b?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 4 }), URLPath("http://example.com/foo/").directory);
+    EXPECT_EQ(Path::Segment({ 19, 4 }), URLPath("http://example.com/foo/?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo.png").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/foo.png?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/").directory);
+    EXPECT_EQ(Path::Segment({ 19, 0 }), URLPath("http://example.com/?query=foo.bar").directory);
+    EXPECT_EQ(Path::Segment({ 18, 0 }), URLPath("http://example.com").directory);
+    EXPECT_EQ(Path::Segment({ 18, 0 }), URLPath("http://example.com?query=foo.bar").directory);
+}
+
+TEST(Path, Extension) {
```

Comment:
I guess it's debatable on what constitutes an extension and what is part of the filename. For a (hypothetical) URL of `mapbox://sources/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7`, would the extension be nothing? or `.mapbox-terrain-v2,mapbox.mapbox-streets-v7`?